### PR TITLE
:whale: docker

### PR DESCRIPTION
**Pull request by guillaumecrespel**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+### BUILD FRONT
+# invoices
+FROM node:9-alpine AS invoices-builder
+WORKDIR /usr/src/app
+COPY ui/ui-invoices/package.json ui/ui-invoices/yarn.lock ./
```
Comment (fabienjuif) [inline Dockerfile]: You can use build arguments:
```Dockerfile
FROM node:9-alpine AS invoices-builder

ARG project

WORKDIR /usr/src/app
COPY ui/${project}/package.json ui/${project}/yarn.lock ./

# ...
```

Comment (fabienjuif) [reply Dockerfile]: And then: ` docker build . --build-arg project=invoices`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
+### BUILD FRONT
+# invoices
+FROM node:9-alpine AS invoices-builder
```
Comment (fabienjuif) [inline Dockerfile]: `as builder` if you want this to be used for both projects

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+WORKDIR /usr/src/app
+COPY ui/ui-invoices/package.json ui/ui-invoices/yarn.lock ./
+RUN yarn install
+COPY ui/ui-invoices/ .
+RUN yarn build
+
+FROM node:9-alpine AS kmille-invoices
```
Comment (fabienjuif) [inline Dockerfile]: I don't think you need the `AS` here since this stage is not used after

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+WORKDIR /usr/src/app
+COPY package.json lerna.json ./
+COPY packages ./packages
+COPY api/api-invoices ./api/api-invoices
+RUN yarn install
+COPY --from=invoices-builder /usr/src/app/build ./api/api-invoices/build
+RUN node api/api-invoices/src/index.js
```
Comment (fabienjuif) [inline Dockerfile]: This is not `RUN` this is `ENTRYPOINT`

- `RUN` is a buildtime command
- `ENTRYPOINT` is a runtime command

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+EXPOSE 3001
```
Comment (fabienjuif) [inline Dockerfile]: This line should be first so it is cached (just after FROM)

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+RUN yarn install
+COPY ui/ui-invoices/ .
+RUN yarn build
+
+FROM node:9-alpine AS kmille-invoices
+WORKDIR /usr/src/app
+COPY package.json lerna.json ./
+COPY packages ./packages
+COPY api/api-invoices ./api/api-invoices
+RUN yarn install
```
Comment (fabienjuif) [inline Dockerfile]: I think this is better to run yarn install in production mode (image size)

**Comment by fabienjuif** [comment]
Maybe we should add a root script (package) json to build project :)

like `yarn build:invoices`
that call a common script `run:build` (like `run:start`)

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# build front
+FROM node:9-alpine as builder
+WORKDIR /usr/src/app
+ARG app
+COPY ui/ui-$app/package.json ./
```
Comment (fabienjuif) [inline Dockerfile]: copy lock

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+RUN yarn install
+COPY ui/ui-$app/ .
+RUN yarn build
+
+FROM node:9-alpine
+WORKDIR /usr/src/app
+COPY package.json lerna.json ./
```
Comment (fabienjuif) [inline Dockerfile]: copy lock

Comment (fabienjuif) [reply Dockerfile]: I don't think you need top copy lerna.json

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+COPY packages ./packages
+ARG app
+COPY api/api-$app ./api/api-$app
+RUN yarn install
+COPY --from=builder /usr/src/app/build ./api/api-$app/build
+EXPOSE 3001
```
Comment (fabienjuif) [inline Dockerfile]: endline

File: docker-compose.template.yml [hunk 10]
```diff
@@ -0,0 +1,61 @@
+version: '3.2'
+services:
+  invoices:
+    image: invoices
+    user: "node"
+    working_dir: /usr/src/app
+    command: "node ./api/api-invoices/src/index.js"
+    restart: always
+    ports:
```
Comment (fabienjuif) [inline docker-compose.template.yml]: you don't need to bind ports since you use traefik

File: docker-compose.template.yml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+      - target: 3001
+        published: 8001
+        protocol: tcp
+        mode: host
+    labels:
+      - "traefik.port=8001"
```
Comment (fabienjuif) [inline docker-compose.template.yml]: 3001

File: docker-compose.template.yml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      - "traefik.frontend.rule=Host:invoices.<DNS>"
+      - "traefik.frontend.entryPoints=https"
+    environment:
+      - NODE_ENV=production
+      - GOOGLE_AUTH_CLIENT_ID=
+      - GOOGLE_AUTH_CLIENT_SECRET=
+      - GOOGLE_AUTH_REDIRECT=http://localhost:3001/api/auth/google/callback
```
Comment (fabienjuif) [inline docker-compose.template.yml]: this is not the good redirect I think

File: docker-compose.template.yml [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        mode: host
+    labels:
+      - "traefik.port=8002"
+      - "traefik.frontend.rule=Host:expenses.<DNS>"
+      - "traefik.frontend.entryPoints=https"
+    environment:
+      - NODE_ENV=production
+      - GOOGLE_AUTH_CLIENT_ID=
+      - GOOGLE_AUTH_CLIENT_SECRET=
+      - GOOGLE_AUTH_REDIRECT=http://localhost:3001/api/auth/google/callback
```
Comment (fabienjuif) [inline docker-compose.template.yml]: redirect

File: docker-compose.template.yml [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+      - target: 3001
+        published: 8002
+        protocol: tcp
+        mode: host
+    labels:
+      - "traefik.port=8002"
+      - "traefik.frontend.rule=Host:expenses.<DNS>"
+      - "traefik.frontend.entryPoints=https"
+    environment:
+      - NODE_ENV=production
```
Comment (fabienjuif) [inline docker-compose.template.yml]: I think you can use a env file to share it between apps

File: docker-compose.template.yml [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+      - ACCOUNTANT_EMAIL_SUBJECT_INVOICES="ALAKARTE – FACTURE DE VENTE"
+  
+  expenses:
+    image: expenses
+    user: "node"
+    working_dir: /usr/src/app
+    command: "node ./api/api-expenses/src/index.js"
+    restart: always
+    ports:
+      - target: 3001
```
Comment (fabienjuif) [inline docker-compose.template.yml]: You don't need it with traefik

File: docker-compose.template.yml [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        published: 8002
+        protocol: tcp
+        mode: host
+    labels:
+      - "traefik.port=8002"
```
Comment (fabienjuif) [inline docker-compose.template.yml]: 3001

File: docker-compose.template.yml (hunk 10 above)
Comment (guillaumecrespel) [reply docker-compose.template.yml]: I don't use traefik on local

Comment (fabienjuif) [reply docker-compose.template.yml]: Your docker-compose is for the production purpose here.
So if you want to test something on local create an other docker-compose.

**Comment by guillaumecrespel** [comment]
- [ ] i think front build doesn't work
- [ ] update ubuntu version of guikai.org

File: build.sh [hunk 17]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+docker build -t expenses --build-arg app=expenses . --rm=true
```
Comment (fabienjuif) [inline build.sh]: see my comment about package json script

File: package.json [hunk 18]
```diff
@@ -10,6 +10,8 @@
     "run:build": "lerna run build --stream --scope",
     "build:invoices": "npm run run:build ui-invoices",
     "build:expenses": "npm run run:build ui-expenses",
+    "docker:invoices": "docker build -t invoices --build-arg app=invoices . --rm=true",
```
Comment (fabienjuif) [inline package.json]: re-read comment please

Comment (fabienjuif) [inline package.json]: Why don't you use `run:docker` ? (look previous line as an exemple)

**Comment by fabienjuif** [comment]
@guillaumecrespel take care of you git configuration (git config --global user.name and git config --global user.email)

File: package.json (hunk 18 above)
Comment (guillaumecrespel) [reply package.json]: how you get param `app` at two emplacements ?

**Comment by fabienjuif** [comment]
What do you still have to do ?

**Pull request closed without merging**